pint_server/models.py

```diff
@@ -59,59 +60,74 @@ def __repr__(self):
 
 
 class ProviderImageBase(PintBase):
-    name = Column(String(255), primary_key=True)
-    state = Column(Enum(ImageState, name=ImageState.__enum_name__))
+    state = Column(Enum(ImageState, name=ImageState.__enum_name__),
+                   nullable=False)
     replacementname = Column(String(255))
-    publishedon = Column(Date, primary_key=True)
+    publishedon = Column(Date, nullable=False)
     deprecatedon = Column(Date)
     deletedon = Column(Date)
     changeinfo = Column(String(255))
 
 
 class ProviderServerBase(PintBase):
-    type = Column(Enum(ServerType, name=ServerType.__enum_name__))
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    type = Column(Enum(ServerType, name=ServerType.__enum_name__),
+                  nullable=False)
     shape = Column(String(10))
     name = Column(String(100))
     # NOTE(gyee): the INET type is specific to PostgreSQL. If in the future
     # we decided to support other vendors, we'll need to update this
     # column type accordingly.
-    ip = Column(postgresql.INET, primary_key=True)
-    region = Column(String(100), primary_key=True)
+    ip = Column(postgresql.INET)
+    region = Column(String(100), nullable=False)
     ipv6 = Column(postgresql.INET)
 
+    @validates("name")
+    def validate_name(self, key, value):
```

Comment:
Just a comment on style - the `@validates("name")` specificies that we're validating the name attribute; the actual validation performed is that it's not blank on update servers.

Consider `def validate_is_not_blank` ... and using the `key` in the condition & error handler.

Reply:
We could re-use that validation for deprecatedon and deletedon on Images, as well.

Reply:
With regard to deprecatedon and deletedon, I'm assuming that you want to relate the blankness check to the image?

As for the validate_name() validator, since the definition of validator methods is table specific, and the internal implementation is table specific, in that we are checking the self.type field to see if it has a specific value, I'm not sure that it would be worth the effort to finagle things to share an implementation with the equivalent for the images table, as it may actually make the code, which is pretty minimal/simple right now, overly complex.

As regards the choice of `validate_name` as the name for this validator method, well it is accurate, apt and succint given that we are validating the name field, and then internally the implemented check is checking for specific conditions that apply to the name field. I'm not sure that the same applies for `validate_is_not_blank` given that the name field can be blank for region servers, which would suggest a name of `validate_name_is_not_blank_for_update_servers` which is getting somewhat unwieldy.

I suppose we could implement a helper function that raises ValueError if the value is blank, that `validate_name` could call... But for that to be generic enough to be shareable between valdators it would need to either take the table, key, value and an exception message fragment as arguments so that it could construct the required exception message, or take the value and exception message itself as arguments... Which would greatly increase the complexity of what is currently relatively simple code, for not a whole lot of benefit maintainability wise.

Reply:
Good point on using the key argument in the exception message - I'll fix that now.